cudamapper/src/index_generator_gpu.cu

```diff
@@ -753,309 +757,413 @@ namespace claragenomics {
-        read_id_to_compressed_minimizers_d.free();
-        CGA_LOG_INFO("Deallocating {} bytes from read_id_to_windows_section_d", read_id_to_windows_section_d.size()*sizeof(decltype(read_id_to_windows_section_d)::value_type));
-        read_id_to_windows_section_d.free();
-    
-        // *** sort minimizers by representation ***
-        // As this is a stable sort and the data was initailly grouper by read_id this means that the minimizers within each representations are sorted by read_id
-        thrust::stable_sort_by_key(thrust::device, representations_compressed_d.data(), representations_compressed_d.data() + total_minimizers, rest_compressed_d.data());
-
-        std::vector<representation_t> representations_compressed_h(total_minimizers);
-        std::vector<ReadidPositionDirection> rest_compressed_h(total_minimizers);
-        CGA_CU_CHECK_ERR(cudaMemcpy(representations_compressed_h.data(),
-                                    representations_compressed_d.data(),
-                                    representations_compressed_h.size()*sizeof(decltype(representations_compressed_h)::value_type),
-                                    cudaMemcpyDeviceToHost
-                                    )
-                         );
-        CGA_CU_CHECK_ERR(cudaMemcpy(rest_compressed_h.data(),
-                                    rest_compressed_d.data(),
-                                    rest_compressed_h.size()*sizeof(decltype(rest_compressed_h)::value_type),
-                                    cudaMemcpyDeviceToHost
-                                    )
-                         );
-
-        // free these arrays as they are not needed anymore
-        CGA_LOG_INFO("Deallocating {} bytes from representations_compressed_d", representations_compressed_d.size()*sizeof(decltype(representations_compressed_d)::value_type));
-        representations_compressed_d.free();
-        CGA_LOG_INFO("Deallocating {} bytes from rest_compressed_d", rest_compressed_d.size()*sizeof(decltype(rest_compressed_d)::value_type));
-        rest_compressed_d.free();
+            all_representation_readid_position_direction.push_back(representation_read_id_position_direction);
 
+            if (parser_status == false) {
+                break;
+            }
+        }
         // *** add the minimizers to the host side index ***
         // minimizers are already sorted by representation -> add all minimizers with the same representation to a vector and then add that vector to the hash table
+        std::vector<std::pair<representation_t, ReadidPositionDirection>> repr_rest_pairs;
+
+        //Lambda function to compare std::pair<representation_t, ReadidPositionDirection>
+        std::function<bool(std::pair<representation_t, ReadidPositionDirection>, std::pair<representation_t, ReadidPositionDirection>)> compare =
```

Comment:
I no longer change it to a function call (templatised `merge_n_sorted_vectors`) so it's implemented like this:


```
        auto compare = [](std::pair<representation_t, ReadidPositionDirection> a,  std::pair<representation_t, ReadidPositionDirection> b){
            return a.first < b.first;
        };
```

It's scope is already limited because it has an empty capture clause `[ ]` and can access only those variable which are local to it. I find this a bit cleaner and I think performance should be the same. What do you think?

Reply:
By scope I meant that the lambda lives until the end of the constructor, not until the end of the merging function. This way you create an lvalue, whereas when you create it inside the function call it's an rvalue which can be moved (not sure if there is performance difference in this case)
I think most people write lambdas inside function calls, but it's almost a matter of taste